jlm/llvm/opt/CommonNodeElimination.cpp

```diff
@@ -0,0 +1,1017 @@
+  {
+    const auto & leaderOutput = *leader.output(i);
+    const auto & followerOutput = *follower.output(i);
+    const auto leaderSet = congruenceSets.getSetFor(leaderOutput);
+    congruenceSets.addFollower(leaderSet, followerOutput);
+  }
+}
+
+/**
+ * Checks if the given \p node has been visited already during the mark phase.
+ * If so, it returns the leader it is congruent with.
+ * All outputs of \p node are congruent with the respective output of the leader.
+ * @param node the node in question.
+ * @param context the current context of the mark phase.
+ * @return the leader of this node's congruence sets, or nullptr if this node has not been marked.
+ */
+[[nodiscard]] const rvsdg::Node *
+tryGetLeaderNode(const rvsdg::Node & node, CommonNodeElimination::Context & context)
+{
+  // Nodes with 0 outputs are never congruent with anything, so let the node be its own leader.
+  if (node.noutputs() == 0)
+    return &node;
+
+  const auto & congruenceSets = context.getCongruenceSets();
+
+  // Check the congruence set of the first output
+  const auto output0Set = congruenceSets.tryGetSetFor(*node.output(0));
+  // If the output has not gotten a congruence set yet, the node has yet to be marked
+  if (output0Set == CongruenceSets::NoCongruenceSet)
+    return nullptr;
+
+  // Structural nodes can have outputs that are congruent with completely different nodes
+  // due to invariant values. Structural nodes are always considered their own leaders.
+  if (dynamic_cast<const rvsdg::StructuralNode *>(&node))
+    return &node;
+
+  // A simple node can only be congruent with other simple nodes
+  const auto & output0Leader = congruenceSets.getLeader(output0Set);
+  const auto & leaderNode = rvsdg::AssertGetOwnerNode<rvsdg::SimpleNode>(output0Leader);
+  JLM_ASSERT(leaderNode.noutputs() == node.noutputs());
+  return &leaderNode;
+}
+
+using TopNodeLeaderList = std::vector<const rvsdg::Node *>;
+
+/**
+ * Function for marking the given top node \p node, marking it congruent with a previously marked
+ * top node if their operations are identical. Otherwise it becomes a leader.
+ * Maintains a separate list of leader top nodes in the region.
+ * @param node the top node to mark.
+ * @param leaders the list of all leader top nodes in the region.
+ * @param context the current context of the marking phase.
+ */
+static void
+markSimpleTopNode(
+    const rvsdg::SimpleNode & node,
+    TopNodeLeaderList & leaders,
+    CommonNodeElimination::Context & context)
+{
+  // If the node has already been marked, check if it is still congruent with its leader
+  if (const auto existingLeaderNode = tryGetLeaderNode(node, context))
+  {
+    // We are our own leader, nothing to check
+    if (existingLeaderNode == &node)
+    {
+      leaders.push_back(&node);
+      return;
+    }
+
+    // Check if we are still congruent with the leader
+    if (checkNodesCongruent(*existingLeaderNode, node, context))
+    {
+      return;
+    }
```

Comment:
Why can it ever happen that we were once congruent with another node, but not any longer?

According to my understanding, this should never happen

Reply:
yup